lib/nerves_hub_link_common/downloader.ex

```diff
@@ -167,8 +178,45 @@ defmodule NervesHubLinkCommon.Downloader do
   # schedules a message to be delivered based on retry args
   @spec reschedule_resume(t()) :: resume_rescheduled()
   defp reschedule_resume(%Downloader{retry_number: retry_number} = state) do
+    # cancel the worst_case_timeout if it was running
+    worst_case_timeout_remaining_ms =
+      if state.worst_case_timeout do
+        Process.cancel_timer(state.worst_case_timeout) || nil
+      end
+
     timer = Process.send_after(self(), :resume, state.retry_args.time_between_retries)
-    %Downloader{state | retry_timeout: timer, retry_number: retry_number + 1}
+
+    %Downloader{
+      state
+      | retry_timeout: timer,
+        retry_number: retry_number + 1,
+        worst_case_timeout_remaining_ms: worst_case_timeout_remaining_ms
+    }
+  end
+
+  @spec schedule_worst_case_timer(t()) :: t()
+  # only calculate worst_case_timeout_remaining_ms is not set
+  defp schedule_worst_case_timer(%Downloader{worst_case_timeout_remaining_ms: nil} = downloader) do
+    # decompose here because in the formatter doesn't like all this being in the head
+    %Downloader{retry_args: retry_config, content_length: content_length} = downloader
+    %RetryConfig{worst_case_download_speed: speed} = retry_config
+
+    # need to extract milliseconds based on a speed in seconds and number of bits
+    ms = round(content_length * 8 / speed * 1000)
+
+    # check to make sure that there's enough data to even conceivably reach this timeout
+    timer =
+      if content_length * 8 >= speed,
+        do: Process.send_after(self(), :worst_case_download_speed_timeout, ms)
```

Comment:
How about setting a minimum timeout of a minute? 

```suggestion
    ms = round(content_length * 8 / speed * 1000) |> max(60_000)

    timer = Process.send_after(self(), :worst_case_download_speed_timeout, ms)
```